include/vcpkg/base/parallel-algorithms.h

```diff
@@ -1,85 +1,168 @@
-        const size_t num_threads = std::max(static_cast<size_t>(1), std::min(thread_count, work_count));
+        F work;
+        size_t work_count;
+        std::atomic<size_t> next_offset;
 
-        std::vector<std::future<void>> workers;
-        workers.reserve(num_threads - 1);
+        WorkCallbackContext(F init_f, size_t work_count) : work(init_f), work_count(work_count), next_offset(0) { }
 
-        for (size_t i = 0; i < num_threads - 1; ++i)
+        // pre: run() is called at most SIZE_MAX - work_count times
+        void run()
         {
-            workers.emplace_back(std::async(std::launch::async | std::launch::deferred, [&work]() { work(); }));
+            for (;;)
+            {
+                auto offset = next_offset.fetch_add(1, std::memory_order_relaxed);
+                if (offset >= work_count)
+                {
+                    return;
+                }
+
+                work(offset);
+            }
         }
-        work();
+    };
 
-        for (auto&& w : workers)
+#if defined(_WIN32)
+    struct PtpWork
+    {
+        PtpWork(_In_ PTP_WORK_CALLBACK pfnwk, _Inout_opt_ PVOID pv, _In_opt_ PTP_CALLBACK_ENVIRON pcbe)
+            : ptp_work(CreateThreadpoolWork(pfnwk, pv, pcbe))
         {
-            w.get();
         }
-    }
+        PtpWork(const PtpWork&) = delete;
+        PtpWork& operator=(const PtpWork&) = delete;
+        ~PtpWork()
+        {
+            if (ptp_work)
+            {
+                ::WaitForThreadpoolWorkCallbacks(ptp_work, TRUE);
+                ::CloseThreadpoolWork(ptp_work);
+            }
+        }
 
-    template<class Container, class F>
-    void parallel_for_each(Container&& c, F cb) noexcept
+        explicit operator bool() { return ptp_work != nullptr; }
+
+        void submit() { ::SubmitThreadpoolWork(ptp_work); }
+
+    private:
+        PTP_WORK ptp_work;
+    };
+
+    template<class F>
+    inline void execute_in_parallel(size_t work_count, F work) noexcept
     {
-        if (c.size() == 0)
+        if (work_count == 0)
         {
             return;
         }
-        if (c.size() == 1)
+
+        if (work_count == 1)
         {
-            cb(c[0]);
-            return;
+            work(size_t{});
         }
 
-        std::atomic_size_t next{0};
-
-        execute_in_parallel(c.size(), [&]() {
-            size_t i = 0;
-            while (i < c.size())
+        WorkCallbackContext<F> context{work, work_count};
+        PtpWork ptp_work([](PTP_CALLBACK_INSTANCE,
+                            void* context,
+                            PTP_WORK) noexcept { static_cast<WorkCallbackContext<F>*>(context)->run(); },
+                         &context,
+                         nullptr);
+        if (ptp_work)
+        {
+            auto max_threads = (std::min)(work_count, static_cast<size_t>(get_concurrency()));
+            max_threads = (std::min)(max_threads, SIZE_MAX - work_count); // to avoid overflow in fetch_add
```

Comment:
```suggestion
            if (SIZE_MAX - work_count > max_threads) Checks::unreachable(VCPKG_LINE_INFO);
```
I don't think trying to be fancy here to support more than 18,446,744,073,709,550,000-ish (or 4,294,966,000-ish) items is worth it.

Also, I think the current line is pessimistic by 1 item/thread -- if `work_count == SIZE_MAX`, 1 thread is permissible.

Reply:
>I don't think trying to be fancy here to support more than 18,446,744,073,709,550,000-ish (or 4,294,966,000-ish) items is worth it.

I think it is worth it, at least as long as 32 bit is not dead. (And arm-linux is still very much a thing, even if we won't get much parallelism there)

>Also, I think the current line is pessimistic by 1 item/thread -- if work_count == SIZE_MAX, 1 thread is permissible.

Sure. (Note that `work_count == SIZE_MAX` is impossible because the `end()` element can't exist in the address space. :) )

Reply:
(Moreover, a dependency from this stuff on Checks:: makes this more complex, not less)